### ISSUE TITLE
[Question] listening for changes on an observed node

### ISSUE THREAD

**Issue by thehabbos007** [lord/anchors#35]
Hey!
Want to start out by saying this is some very cool work!
I've been playing around with the library a little, and came to a point where I wanted to listen for changes when an observed node has changed its inner value.

I wanted to know if there is a mechanism for having a sort of "on change" callback for observed nodes? I tried implementing it myself with closures within the `Var` struct, but it seemed to bring problems - and im not sure how to implement such a feature the way map and cutoff are implemented :)

Happy new year!

**Comment by lord** [comment]
Hey, sorry, just saw this! Would definitely be a great feature to add, thanks for the suggestion! I'll try to get to it in a couple days.

**Comment by thehabbos007** [comment]
It's all good! I was interested in helping out too, but I don't know if i understand the architecture well enough :)

Thank you a lot!

**Comment by lord** [comment]
If you wanted to try to help hack on it, I think maybe the best way to solve it is to refactor the current mark_observed and mark_unobserved functions into a kind of Observer object, and maybe allow subscribing to changes on that object? Kind of a vague suggestion, sorry.

Glad somebody out there is actually using this library, and sorry for the poor documentation!

**Comment by thehabbos007** [comment]
> If you wanted to try to help hack on it, I think maybe the best way to solve it is to refactor the current mark_observed and mark_unobserved functions into a kind of Observer object

I'll try to play around with this idea. Would be a nice learning experience for myself as well :)

**Comment by lord** [comment]
I think you can actually do this already without any changes to Anchors, although perhaps it can be made more ergonomic? Just added this as `example/update_callback.rs`:

```rs
use anchors::expert::{AnchorExt, Var};
use anchors::singlethread::*;
use std::cell::RefCell;

fn main() {
    let mut engine = Engine::new();
    let (cat_count, set_cat_count) = Var::new(1);
    let (dog_count, set_dog_count) = Var::new(1);
    let (fish_count, set_fish_count) = Var::new(1);
    let total_mammals = (&cat_count, &dog_count).map(|cats, dogs| cats + dogs);
    let total_animals = (&total_mammals, &fish_count).map(|mammals, fish| mammals + fish);
    let mammal_callback = total_mammals.map(|total_mammals| println!("mammals updated: {:?}", total_mammals));
    let animal_callback = total_animals.map(|total_animals| println!("animals updated: {:?}", total_animals));
    engine.mark_observed(&mammal_callback);
    engine.mark_observed(&animal_callback);

    println!("stabilizing...");
    engine.stabilize();

    set_cat_count.set(2);
    set_dog_count.set(2);
    println!("stabilizing...");
    engine.stabilize();

    set_fish_count.set(2);
    println!("stabilizing...");
    engine.stabilize();
}
```

**Comment by thehabbos007** [comment]
Ah, I see, of course using map would be a way to accomplish this. 
What I had played around with was adding a callback function to the var `new` method and called it `new_observed` having that take in a callback closure which was added onto the returned `Var`. I'd then call down to that callback in `poll_updated`. Although it does work, it required a lot of finicky type declarations to make the compiler happy, not really a solution I liked. 

This is more sensible :)
